### PR TITLE
Skip KubernetesPodOperator task when it returns a provided exit code

### PR DESCRIPTION
**Pull request by hussein-awala**
Skip `KubernetesPodOperator` task when the pod container returns the exit code `skip_exit_code` provided as an argument.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
@jedcunningham - maybe you can take a look as well especially at the exit-code retrieval part.

**Comment by hussein-awala** [comment]
This needs some tests for deferred mode, I convert it to draft

File: airflow/providers/cncf/kubernetes/operators/pod.py [hunk 1]
```diff
@@ -670,6 +675,24 @@ def cleanup(self, pod: k8s.V1Pod, remote_pod: k8s.V1Pod):
 
             error_message = get_container_termination_message(remote_pod, self.base_container_name)
             error_message = "\n" + error_message if error_message else ""
+            if self.skip_exit_code is not None:
+                container_statuses = (
+                    remote_pod.status.container_statuses if remote_pod and remote_pod.status else None
+                )
+                base_container_status = next(
+                    (x for x in container_statuses if x.name == self.base_container_name), None
```
Comment (jedcunningham) [inline airflow/providers/cncf/kubernetes/operators/pod.py:683]: If `container_statuses` is None, won't this complain that None isn't iterable?

File: tests/providers/cncf/kubernetes/operators/test_pod.py [hunk 2]
```diff
@@ -1052,6 +1052,35 @@ def test_task_id_as_name_dag_id_is_ignored(self):
...
+            ({"skip_exit_code": 100}, 100, AirflowSkipException),
+            ({"skip_exit_code": 100}, 101, AirflowException),
+            ({"skip_exit_code": None}, 100, AirflowException),
+        ],
+    )
+    @patch(f"{POD_MANAGER_CLASS}.await_pod_completion")
+    def test_task_skip_when_pod_exit_with_certain_code(
+        self, remote_pod, extra_kwargs, actual_exit_code, expected_exc
+    ):
+        """ """
```
Comment (jedcunningham) [inline tests/providers/cncf/kubernetes/operators/test_pod.py]: Empty doc string

File: tests/providers/cncf/kubernetes/operators/test_pod.py [hunk 3]
```diff
@@ -1217,6 +1246,70 @@ def test_async_create_pod_should_throw_exception(self, mocked_hook, mocked_clean
...
+    def test_async_create_pod_with_skip_exit_code_should_skip(
+        self,
+        mocked_hook,
+        extra_kwargs,
+        actual_exit_code,
+        expected_exc,
+        pod_status,
+        event_status,
+    ):
+        """Tests that an AirflowSkipException is raised when the pod exit with the skip_exit_code"""
```
Comment (jedcunningham) [inline tests/providers/cncf/kubernetes/operators/test_pod.py]:
```suggestion
        """Tests that an AirflowSkipException is raised when the container exits with the skip_exit_code"""
```

nit

File: airflow/providers/cncf/kubernetes/operators/pod.py (hunk 1 above)
Comment (hussein-awala) [reply airflow/providers/cncf/kubernetes/operators/pod.py:683]: :+1:  I added `or []` for container_statuses, this can fix the problem

**Comment by hussein-awala** [comment]
@jedcunningham can you please review the last changes?

**Review by eladkal — APPROVED** [review]

**Pull request merged**